library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -2154,17 +1856,26 @@ impl<T, A: Allocator> VecDeque<T, A> {
     #[inline]
     #[stable(feature = "append", since = "1.4.0")]
     pub fn append(&mut self, other: &mut Self) {
-        self.reserve(other.len());
+        if T::IS_ZST {
+            self.len += other.len;
+            other.len = 0;
+            other.head = 0;
+            return;
+        }
+
+        self.reserve(other.len);
         unsafe {
             let (left, right) = other.as_slices();
-            self.copy_slice(self.head, left);
-            self.copy_slice(self.wrap_add(self.head, left.len()), right);
+            self.copy_slice(self.wrap_idx(self.len), left);
+            // no overflow, because self.capacity() >= old_cap + left.len() >= self.len + left.len()
+            self.copy_slice(self.wrap_idx(self.len + left.len()), right);
         }
         // SAFETY: Update pointers after copying to avoid leaving doppelganger
         // in case of panics.
-        self.head = self.wrap_add(self.head, other.len());
+        self.len += other.len;
         // Silently drop values in `other`.
-        other.tail = other.head;
+        other.len = 0;
+        other.head = 0;
```

Comment:
less about this specific line but more about the pattern:

How much is it worth actually trying to remember to zero out `head` in all the places?  Since we *could* just not bother without violating invariants, AFAICT, and I wouldn't really want to do an exhaustive survey to *require* that `len == 0` implies `head == 0`, maybe it'd be better to just not bother?

Reply:
I'm not sure, I'll have to look what code actually depends on zeroing out the `head`. I guess in my mind it was just more intuitive, that if you somehow empty the deque, then elements added afterwards fill up the buffer contiguously.